File: Backend/schemas/kanban.py
# schemas/kanban.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field
from uuid import UUID


# ---------------------------------------------------------------------------
# Boards
# ---------------------------------------------------------------------------
class BoardCreate(BaseModel):
    project_id: int
    sprint_id: Optional[int] = None
    name: str
    description: Optional[str] = None


class BoardUpdate(BaseModel):
    sprint_id: Optional[int] = None
    name: Optional[str] = None
    description: Optional[str] = None


class ColumnResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    board_id: int
    name: str
    status: str
    wip_limit: Optional[int]
    order: float
    created_at: datetime


class BoardResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    sprint_id: Optional[int]
    name: str
    description: Optional[str]
    created_at: datetime
    updated_at: Optional[datetime]


class BoardDetailResponse(BoardResponse):
    columns: list["ColumnDetailResponse"]
    tasks: list["TaskResponse"]


# ---------------------------------------------------------------------------
# Columns
# ---------------------------------------------------------------------------
class ColumnCreate(BaseModel):
    name: str
    status: str = Field(description="One of: todo, in_progress, in_review, done, blocked")
    wip_limit: Optional[int] = Field(default=None, ge=0)
    order: Optional[float] = None


class ColumnUpdate(BaseModel):
    name: Optional[str] = None
    status: Optional[str] = None
    wip_limit: Optional[int] = Field(default=None, ge=0)
    order: Optional[float] = None


class ColumnDetailResponse(ColumnResponse):
    tasks: list["TaskResponse"]


# ---------------------------------------------------------------------------
# Tasks
# ---------------------------------------------------------------------------
class TaskCreate(BaseModel):
    title: str
    description: Optional[str] = None
    column_id: Optional[int] = None
    task_id: Optional[int] = None
    priority: str = "medium"
    assignee_id: Optional[UUID] = None
    story_points: Optional[float] = None
    time_estimate: Optional[float] = None
    time_spent: Optional[float] = 0
    due_date: Optional[datetime] = None
    order: Optional[float] = None
    is_blocked: Optional[bool] = False
    blocked_reason: Optional[str] = None
    tags: Optional[list[str]] = None


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    column_id: Optional[int] = None
    priority: Optional[str] = None
    assignee_id: Optional[UUID] = None
    story_points: Optional[float] = None
    time_estimate: Optional[float] = None
    time_spent: Optional[float] = None
    due_date: Optional[datetime] = None
    order: Optional[float] = None
    is_blocked: Optional[bool] = None
    blocked_reason: Optional[str] = None
    tags: Optional[list[str]] = None


class TaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    board_id: int
    column_id: int
    task_id: Optional[int]
    title: str
    description: Optional[str]
    priority: str
    assignee_id: Optional[UUID]
    assignee_name: Optional[str] = None
    assignee_avatar: Optional[str] = None
    story_points: Optional[float]
    time_estimate: Optional[float]
    time_spent: Optional[float]
    due_date: Optional[datetime]
    order: float
    status: str
    is_blocked: bool
    blocked_reason: Optional[str]
    is_completed: bool
    tags: Optional[list[str]]
    completed_at: Optional[datetime]
    created_at: datetime
    updated_at: Optional[datetime]
    comment_count: int = 0


# ---------------------------------------------------------------------------
# Movement / reorder / bulk
# ---------------------------------------------------------------------------
class TaskMoveRequest(BaseModel):
    column_id: int
    order: Optional[float] = None


class BulkMoveRequest(BaseModel):
    task_ids: list[int]
    column_id: int
    order: Optional[float] = None


class ReorderRequest(BaseModel):
    task_ids: list[int]


class BulkAssignRequest(BaseModel):
    task_ids: list[int]
    assignee_id: Optional[UUID] = None


class BulkPriorityRequest(BaseModel):
    task_ids: list[int]
    priority: str


class TaskBlockRequest(BaseModel):
    reason: Optional[str] = None


# ---------------------------------------------------------------------------
# Search / stats / workload
# ---------------------------------------------------------------------------
class TaskSearchResponse(BaseModel):
    items: list[TaskResponse]
    total: int
    page: int
    page_size: int


class BoardStatsResponse(BaseModel):
    board_id: int
    total: int
    by_status: dict[str, int]
    todo: int
    in_progress: int
    in_review: int
    done: int
    blocked: int
    completed: int
    completion_percentage: float
    total_story_points: float
    completed_story_points: float


class ColumnStatsResponse(BaseModel):
    column_id: int
    board_id: int
    name: str
    status: str
    wip_limit: Optional[int]
    total_tasks: int
    at_limit: bool
    over_limit: bool
    by_priority: dict[str, int]
    blocked_count: int
    avg_story_points: Optional[float]
    total_story_points: float
    total_time_estimate: float
    total_time_spent: float


class WorkloadResponse(BaseModel):
    board_id: int
    user_id: UUID
    total_tasks: int
    completed_tasks: int
    blocked_tasks: int
    in_progress_tasks: int
    total_story_points: float
    completed_story_points: float
    total_time_spent: float
    tasks: list[TaskResponse]


BoardDetailResponse.model_rebuild()
ColumnDetailResponse.model_rebuild()

File: frontend/src/types/kanban.ts
export type KanbanPriority = "critical" | "high" | "medium" | "low";
export type KanbanStatus =
  | "todo"
  | "in_progress"
  | "in_review"
  | "done"
  | "blocked";

export interface KanbanColumn {
  id: number;
  board_id: number;
  name: string;
  status: KanbanStatus;
  wip_limit: number | null;
  order: number;
  created_at: string;
}

export interface KanbanTask {
  id: number;
  board_id: number;
  column_id: number;
  task_id: number | null;
  title: string;
  description: string | null;
  priority: KanbanPriority;
  assignee_id: string | null;
  assignee_name: string | null;
  assignee_avatar: string | null;
  story_points: number | null;
  time_estimate: number | null;
  time_spent: number | null;
  due_date: string | null;
  order: number;
  status: KanbanStatus;
  is_blocked: boolean;
  blocked_reason: string | null;
  is_completed: boolean;
  tags: string[] | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string | null;
  comment_count: number;
}

export interface KanbanBoard {
  id: number;
  project_id: number;
  sprint_id: number | null;
  name: string;
  description: string | null;
  created_at: string;
  updated_at: string | null;
}

export interface KanbanBoardDetail extends KanbanBoard {
  columns: KanbanColumnDetail[];
  tasks: KanbanTask[];
}

export interface KanbanColumnDetail extends KanbanColumn {
  tasks: KanbanTask[];
}

export interface BoardStats {
  board_id: number;
  total: number;
  by_status: Record<string, number>;
  todo: number;
  in_progress: number;
  in_review: number;
  done: number;
  blocked: number;
  completed: number;
  completion_percentage: number;
  total_story_points: number;
  completed_story_points: number;
}

export interface ColumnStats {
  column_id: number;
  board_id: number;
  name: string;
  status: KanbanStatus;
  wip_limit: number | null;
  total_tasks: number;
  at_limit: boolean;
  over_limit: boolean;
  by_priority: Record<string, number>;
  blocked_count: number;
  avg_story_points: number | null;
  total_story_points: number;
  total_time_estimate: number;
  total_time_spent: number;
}

export interface TaskSearchResult {
  items: KanbanTask[];
  total: number;
  page: number;
  page_size: number;
}

export interface Workload {
  board_id: number;
  user_id: string;
  total_tasks: number;
  completed_tasks: number;
  blocked_tasks: number;
  in_progress_tasks: number;
  total_story_points: number;
  completed_story_points: number;
  total_time_spent: number;
  tasks: KanbanTask[];
}

export interface KanbanUser {
  id: string;
  name: string;
  email: string;
  avatar: string | null;
}

export interface TaskCreatePayload {
  title: string;
  description?: string | null;
  column_id?: number | null;
  task_id?: number | null;
  priority?: KanbanPriority;
  assignee_id?: string | null;
  story_points?: number | null;
  time_estimate?: number | null;
  time_spent?: number | null;
  due_date?: string | null;
  order?: number | null;
  is_blocked?: boolean;
  blocked_reason?: string | null;
  tags?: string[] | null;
}

export interface TaskUpdatePayload {
  title?: string;
  description?: string | null;
  column_id?: number | null;
  priority?: KanbanPriority;
  assignee_id?: string | null;
  story_points?: number | null;
  time_estimate?: number | null;
  time_spent?: number | null;
  due_date?: string | null;
  order?: number | null;
  is_blocked?: boolean;
  blocked_reason?: string | null;
  tags?: string[] | null;
}

export interface BoardCreatePayload {
  project_id: number;
  sprint_id?: number | null;
  name: string;
  description?: string | null;
}

export interface ColumnCreatePayload {
  name: string;
  status: KanbanStatus;
  wip_limit?: number | null;
  order?: number | null;
}

export interface ColumnUpdatePayload {
  name?: string;
  status?: KanbanStatus;
  wip_limit?: number | null;
  order?: number | null;
}

export const PRIORITY_LABELS: Record<KanbanPriority, string> = {
  critical: "Critical",
  high: "High",
  medium: "Medium",
  low: "Low",
};

export const STATUS_LABELS: Record<KanbanStatus, string> = {
  todo: "Todo",
  in_progress: "In Progress",
  in_review: "In Review",
  done: "Done",
  blocked: "Blocked",
};